<figure>
  <IMG SRC="https://upload.wikimedia.org/wikipedia/commons/thumb/d/d5/Fachhochschule_Südwestfalen_20xx_logo.svg/320px-Fachhochschule_Südwestfalen_20xx_logo.svg.png" WIDTH=250 ALIGN="right">
</figure>

# Einführung in die Programmierung
### Winterersemester 2025/25
Prof. Dr. Stefan Goetze

# Grundlegende Datentypen

Python ist eine dynamisch typisierte Sprache. Das bedeutet, dass Sie nicht im Voraus festlegen müssen, welche Art von Daten Sie in einer Variable speichern möchten. Dennoch gibt es einige grundlegende Datentypen, mit denen wir uns bei der Verwendung der Sprache vertraut machen müssen. Die ersten Datentypen ähneln denen anderer Sprachen (wie C/C++ und Fortran): Gleitkommazahlen, Ganzzahlen und Zeichenfolgen.

## Fließkommazahlen


Gleitkommazahlen sind für die Informatik von entscheidender Bedeutung. Für diejenigen, die sich noch weitere Deteils dazu ansehen wollen: Eine hervorragende Einführung in die Gleitkommazahlen und ihre Grenzen bietet D. Goldberg: [What every computer scientist should know about floating-point
arithmetic](http://dl.acm.org/citation.cfm?id=103163) by
D. Goldberg.

Die nächsten Datentypen sind Container. Im Gegensatz zu einigen anderen Sprachen sind diese in Python integriert und erleichtern die Durchführung komplexer Operationen. Wir werden uns diese später genauer ansehen.

Einige Beispiele stammen aus dem Python-Tutorial:
http://docs.python.org/3/tutorial/

## Ganzzahlen

Ganzzahlen sind Zahlen ohne Dezimalpunkt. Sie können positiv oder negativ sein. Die meisten Programmiersprachen verwenden einen begrenzten Speicher für die Speicherung einzelner Ganzzahlen. Python erweitert den Speicher jedoch bei Bedarf, um große Ganzzahlen zu speichern.

Die grundlegenden Operatoren „+“, „-“, „*“ und „/“ funktionieren mit Ganzzahlen.

In [62]:
2+2+3

7

In [63]:
2*-4

-8


Die *Ganzzahldivision* ist ein Punkt, in dem sich Python und andere Programmiersprachen unterscheiden.

In Python ergibt die Division zweier Ganzzahlen eine Gleitkommazahl. In C/C++/Fortran ergibt die Division zweier Ganzzahlen eine Ganzzahl, also `1/2 = 0`.

In [64]:
1/2

0.5

Um ein ganzzahliges Ergebnis zu erhalten, können wir den Operator `//` verwenden.

In [65]:
1//2

0

Python ist eine *dynamisch typisierte* Sprache. Das bedeutet, dass wir den Datentyp einer Variable vor der Initialisierung nicht deklarieren müssen.

Hier erstellen wir eine Variable (eine beschreibende Bezeichnung, die auf bestimmte Daten verweisen kann). Der Operator „=“ weist einer Variablen einen Wert zu.

In [66]:
a = 1
b = 2

Funktionen arbeiten mit Variablen und geben ein Ergebnis zurück. Teilweise werden und Berechnung in Jupyter direkt zurückgegeben, wir können aber in jedem Fall den `print()` Befehl nutzen um eine Ausgabe auf dem Bildschirm zu erhalten.

In [67]:
a + b

3

In [68]:
a * b

2

Beachten Sie, dass bei Variablennamen die Groß- und Kleinschreibung beachtet wird. Daher sind `a` und `A` unterschiedlich.

In [69]:
A = 2048

In [70]:
print(a, A)

1 2048


Im folgenden Codeblock initialisieren wir $3$ Variablen, alle auf „$0$“, aber es handelt sich immer noch um unterschiedliche Variablen, sodass wir eine ändern können, ohne die anderen zu beeinflussen.

In [71]:
x = y = z = 0

In [72]:
print(x, y, z)

0 0 0


In [73]:
z = 1

In [74]:
z

1

Python bietet integrierte Hilfe (und Jupyter/iPython sogar noch mehr).

Versuchen Sie Folgendes:
```
help(x)
```

Alternativ können Sie Folgendes versuchen:
```
x?
```

(Funktioniert nur in Jupyter)

Eine weitere hilfreiche Funktion, `type()`, gibt den Datentyp einer Variablen zurück

In [75]:
type(x)

int

In Sprachen wie Fortran und C geben Sie die Speichergröße an, die ein Integer belegen kann (normalerweise $2$ oder $4$ Bytes). Dies schränkt die maximale darstellbare Integer-Größe ein. Python passt die Größe des Integers an, damit es nicht zu einem *Überlauf* kommt.

In [76]:
a = 12345678901234567890123456789012345123456789012345678901234567890
print(a)
print(a.bit_length())
print(type(a))

12345678901234567890123456789012345123456789012345678901234567890
213
<class 'int'>


Hinweis: In Python 3 (das wir verwenden) gibt es keinen Unterschied zwischen den Datentypen `int` und `long`, da diese Typen in Python 3 zu einem einzigen int-Typ zusammengeführt wurden. Früher (in Python 2) war `int` für einen begrenzten Wertebereich und `long` für ganze Zahlen beliebiger Größe zuständig, aber jetzt kann ein einzelner 'int'-Typ auch sehr große Zahlen speichern. 

Der Typ `long` existiert in Python 3 nicht mehr, da seine Funktionalität vollständig in `int` integriert wurde. 

## Fließkommazahlen 

Beim Arbeiten mit Gleitkommazahlen und Ganzzahlen wird das Ergebnis in eine Gleitkommazahl umgewandelt.

Man erkennt Fließkommazahlen am **Komma** (`.` in Python).

In [77]:
1. + 2

3.0

Beachten Sie jedoch den speziellen Ganzzahldivisionsoperator `//`.

In [78]:
1.//2

0.0

Nicht jede Zahl lässt sich in Gleitkommazahlen darstellen. Da es unendlich viele reelle Zahlen zwischen zwei beliebigen Grenzen gibt, der Speicherplatz aber begrenzt ist, müssen wir auf einem Computer Zahlen approximieren. Es gibt einen IEEE-Standard für Gleitkommazahlen, dem nahezu alle Sprachen und Prozessoren folgen.

Das bedeutet zweierlei:

* Nicht jede reelle Zahl lässt sich in Gleitkommazahlen exakt darstellen.
* Zahlen haben eine begrenzte Genauigkeit – unterhalb dieser verlieren wir den Überblick über Unterschiede (dies wird üblicherweise als Rundungsfehler bezeichnet).

Nehmen wir beispielsweise den folgenden Ausdruck:

In [79]:
0.3/0.1 - 3

-4.440892098500626e-16

Als weiteres Beispiel: Die Zahl $0,1$ kann auf einem Computer nicht exakt dargestellt werden. In unserem Ausdruck verwenden wir einen Formatbezeichner (den Teil innerhalb der `{}`), um eine genauere Darstellung zu erreichen:

In [80]:
a = 0.1
print("{:30.20}".format(a))

        0.10000000000000000555


Wir können Python nach den die Grenzen für Gleitkommazahlen fragen:

In [81]:
import sys
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

Dies bedeutet, dass wir nur Zahlen zwischen $2,2250738585072014 \cdot 10^308$ und $1,7976931348623157 \cdot 10^308$ speichern können.

Wir sehen außerdem, dass die Genauigkeit $2,220446049250313 \cdot 10^-16$ beträgt (dies wird allgemein als _Maschinen-Epsilon_ bezeichnet). Um dies zu verdeutlichen, addieren wir eine kleine Zahl zu 1,0. Wir verwenden den Gleichheitsoperator (`==`), um zu prüfen, ob zwei Zahlen gleich sind:

**Aufgabe:**

Definieren Sie zwei Variablen: a = 1 und e = 10^{-16}.

Definieren Sie nun eine dritte Variable: b = a + e.

Wir können den Python-Operator `==` verwenden, um auf Gleichheit zu prüfen. Was erwarten Sie von `b == a`? Führen Sie den Operator aus und prüfen Sie, ob er mit Ihrer Vermutung übereinstimmt.


## Module

Die Python-Kernsprache wird durch eine Standardbibliothek erweitert, die zusätzliche Funktionen bietet. Diese zusätzlichen Komponenten liegen in Form von Modulen vor, die wir in unsere Python-Sitzung (oder unser Programm) importieren können.

Das Modul `math` bietet Funktionen für grundlegende mathematische Operationen und Konstanten (für komplexe Zahlen gibt es ein separates Modul `cmath`).

In Python importieren Sie ein Modul. Die Funktionen werden dann in einem separaten Namespace definiert – einem separaten Bereich, der Namen, Variablen usw. definiert. Eine Variable in einem Namespace kann denselben Namen wie eine Variable in einem anderen Namespace haben, ohne dass es zu Konflikten kommt. Verwenden Sie den Operator „`.`“ für den Zugriff auf ein Mitglied eines Namespace.

Wenn Sie Daten in den Python-Interpreter, hier im Jupyter-Notebook oder in ein Skript eingeben, befindet sich der Inhalt standardmäßig in einem eigenen Standard-Namespace. Sie müssen keiner der Variablen einen Namespace-Indikator voranstellen.

In [86]:
import math

In [87]:
math.pi

3.141592653589793

Dies unterscheidet sich von jeder Variable `pi`, die wir ggf. selbst definieren.

In [88]:
pi = 3

In [89]:
print(pi, math.pi)

3 3.141592653589793


Beachten Sie hier, dass `pi` und `math.pi` sich voneinander unterscheiden – sie befinden sich in unterschiedlichen Namespaces.

## Gleitkommaoperationen

Die gleichen Operatoren, `+`, `-`, `*`, `/` funktionieren wie üblich für Gleitkommazahlen. Um eine Zahl zu potenzieren, verwenden wir den Operator `**` (dies ist dasselbe wie in Fortran).

In [92]:
R = 2.0

In [93]:
math.pi * R**2

12.566370614359172

Die Operatorpriorität entspricht der der meisten Sprachen. Siehe

https://docs.python.org/3/reference/expressions.html#operator-precedence

In der Reihenfolge der Priorität:
* Alles in `()` - Klammern können natürlich verwendet werden, um die Priorität zu überschreiben.
* Slicing, Aufrufe, Indizes
* Potenzierung (`**`)
* `+x`, `-x`, `~x`
* `*`, `@`, `/`, `//`, `%`
* `+`, `-`

(Danach folgen bitweise Operationen und Vergleiche.)

**Augfabe**

Betrachten Sie die folgenden Ausdrücke. Überlegen Sie anhand der Prioritäten, welcher Wert sich ergibt, und probieren Sie ihn anschließend in der Zelle unten aus, um zu sehen, ob Sie richtig lagen.

* `1 + 3*2**2`
* `1 + (3*2)**2`
* `2**3**2`

In [ ]:
## Your code here

Das Mathematikmodul bietet viele der gängigen mathematischen Funktionen, die wir möglicherweise verwenden möchten.

Für die trigonometrischen Funktionen wird erwartet, dass das Funktionsargument in Radianten angegeben wird. Mit `math.radians()` können Sie Grad ins Radianmaß umrechnen, z. B.:

In [94]:
math.cos(math.radians(45))

0.7071067811865476

Beachten Sie, dass wir in dieser Anweisung die Ausgabe einer Funktion (`math.radians()`) in eine zweite Funktion (`math.cos()`) *einspeisen*.

Im Zweifelsfall finden Sie hier Hilfe zum Entdecken aller Funktionen eines Moduls:

In [95]:
help(math.sin)


Help on built-in function sin in module math:

sin(x, /)
    Return the sine of x (measured in radians).



## Zeichenketten (strings)

Python ist es egal, ob Sie einfache oder doppelte Anführungszeichen für Zeichenfolgen verwenden:

In [96]:
a = "this is my string"
b = 'another string'

In [97]:
print(a)
print(b)

this is my string
another string


Many of the usual mathematical operators are defined for strings as well.  For example to concatenate or duplicate:

In [98]:
a + b

'this is my stringanother string'

In [99]:
a + ". " + b

'this is my string. another string'

In [100]:
a * 2

'this is my stringthis is my string'

Es gibt verschiedene Escape-Codes, die in Zeichenfolgen interpretiert werden. Diese beginnen mit einem Backslash, `\`. Beispielsweise können Sie `\n` für eine neue Zeile verwenden.

In [101]:
a = a + "\n" + a 
print(a)

this is my string
this is my string


**Aufgabe**

Mit der Funktion `input()` können Sie den Benutzer um Eingaben bitten.

* Verwenden Sie `help(input)`, um die Funktionsweise zu testen.
* Schreiben Sie Code, um Eingaben anzufordern und das Ergebnis in einer Variable zu speichern. `input()` gibt einen String zurück.

* Verwenden Sie die Funktion `float()`, um eine eingegebene Zahl in eine Gleitkommavariable umzuwandeln.
* Überprüfen Sie mit der Funktion `type()`, ob die Konvertierung funktioniert hat.

In [ ]:
## Your code here ##

""" kann mehrzeilige Zeichenfolgen einschließen. Dies ist nützlich für Docstrings am Anfang von Funktionen (mehr dazu später...)

In [102]:
c = """
Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor 
incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis 
nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore 
eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt 
in culpa qui officia deserunt mollit anim id est laborum."""

In [103]:
print(c)


Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor 
incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis 
nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. 
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore 
eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt 
in culpa qui officia deserunt mollit anim id est laborum.


Ein **Raw-String** ersetzt keine Escape-Sequenzen (wie `\n`). Setzen Sie einfach ein `r` vor das erste Anführungszeichen:

In [104]:
d = r"this is a raw string\n"
d

'this is a raw string\\n'

**Slicing** dient dem Zugriff auf einen Teil eines Strings.

Für Fortran-Anfänger mag das Slicing eines Strings etwas kontraintuitiv erscheinen. Der Trick besteht darin, sich den Index als die linke Kante eines Zeichens im String vorzustellen. Dasselbe gilt auch für spätere Arrays.

Beachten Sie außerdem, dass Python (wie C) eine $0$-basierte Indizierung verwendet.

Negative Indizes zählen von rechts.

In [105]:
a = "this is my string"
print(a)
print(a[5:7])
print(a[0])
print(d)
print(d[-2])

this is my string
is
t
this is a raw string\n
\


**Aufgabe**

Strings verfügen über zahlreiche Methoden (Funktionen, die mit einem bestimmten Datentyp, in diesem Fall Strings, umgehen können). Eine nützliche Methode ist `.find()`. Für den String `a` gibt `a.find(s)` den Index des ersten Vorkommens von `s` zurück.

Für den String `c` oben suchen wir den ersten `.` (identifizieren den ersten vollständigen Satz) und geben mit diesem Ergebnis nur den ersten Satz in `c` aus.


In [ ]:
## Your code here ##

Es gibt auch eine Reihe von Methoden und Funktionen, die mit Zeichenfolgen arbeiten. Hier sind einige Beispiele:

In [106]:
print(a.replace("this", "that"))
print(len(a))
print(a.strip())    # Also notice that strip removes the \n
print(a.strip()[-1])

that is my string
17
this is my string
g


Beachten Sie, dass sich unsere ursprüngliche Zeichenfolge `a` nicht geändert hat. In Python sind Zeichenfolgen *unveränderlich*. Operationen an Zeichenfolgen geben eine neue Zeichenfolge zurück.

In [107]:
a

'this is my string'

In [108]:
type(a)

str

Wir können beim Drucken Zeichenfolgen formatieren, um Mengen an bestimmten Stellen in der Zeichenfolge einzufügen. Ein `{}` dient als Platzhalter für eine Menge und wird mit der Methode `.format()` ersetzt:

In [109]:
a = 1
b = 2.0
c = "test"
print("a = {}; b = {}; c = {}".format(a, b, c))

a = 1; b = 2.0; c = test


Aber die modernere Art, dies zu tun, ist die Verwendung von *f-strings*. (Beachten Sie das `f` vor dem einleitenden `"`)

In [110]:
print(f"a = {a}; b = {b}; c = {c}")

a = 1; b = 2.0; c = test


Hier noch zwei weitere Möglichkeiten (Es bleibt Ihnen überlassen, Ihren **Favoriten** zu finden):

In [112]:
print("a =" ,  a, "; b =", b, "; c =", c)

a = 1 ; b = 2.0 ; c = test


In [114]:
print("a = " + str(a) + "; b = " + str(b) + "; c = " + str(c))

a = 1; b = 2.0; c = test


## Credits

Beiträge zu diesem Notebook: Diese Notebook basiert auf dem Gooele Notebook [Data Types](https://colab.research.google.com/github/sbu-python-class/python-science/blob/main/content/01-python/w1-python-datatypes.ipynb#scrollTo=GyPq679SKgY6), Anpassung und Erweiterung: [Stefan Goetze](https://github.com/Stefan-Goe)